[PyTorch Numeric Suite] Add module output comparison

hx89 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36701 [PyTorch Numeric Suite] Add module output comparison**

Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

Commit:
[PyTorch Numeric Suite] Add module output comparison

Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -178,3 +214,70 @@ def compare_model_stub(float_model, q_model, module_swap_list, data, Logger=Logg
+        and each entry being a dictionary with two keys 'float' and 'quantized',
+        containing the matching float and quantized activations
+    """
+    if white_list is None:
```
Suggest creating a default white_list in https://github.com/pytorch/pytorch/blob/master/torch/quantization/default_mappings.py and using that as a default argument to compare_model_outputs

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -178,3 +214,70 @@ def compare_model_stub(float_model, q_model, module_swap_list, data, Logger=Logg
+        activation=RecordingObserver, weight=None
+    )
+    float_model.qconfig = qconfig_debug
+    prepare(float_model, inplace=True)
```
Also, wondering if it would be better to have two modular APIs:
``` def prepare_model_outputs() and def get_matching_activations()``` that do the operations prior and after  lines 276 and 277 respectively

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -14,61 +16,93 @@ class EagerModeNumericSuiteTest(QuantizationTestCase):
+            ob_dict = compare_model_stub(float_model, q_model, module_swap_list, data)
+            self.assertEqual(len(ob_dict), 1)
+            for k, v in ob_dict.items():
+                self.assertTrue(v["float"].shape == v["quantized"].shape)
```
In addition, can we check if the keys in the observer dict match what is expected? Are we obtaining the activations at all the points in the model?

hx89 replied:
We will get the activations of the modules that are in the white list, may not be all the points in the model, but should cover most.

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -3,6 +3,41 @@
+_INCLUDE_QCONFIG_PROPAGATE_LIST = {nn.Sequential}
+
+
+def compute_error(x, y, error_type="l2"):
```
Is this function used in the code?

hx89 replied:
It's not used in the code, but for user to use when they get the dict from compare APIs to compute the error.

raghuramank100 replied:
Please remove it as it is close to torch.norm, we can add it if needed later.

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -178,3 +214,70 @@ def compare_model_stub(float_model, q_model, module_swap_list, data, Logger=Logg
+    float_model.qconfig = qconfig_debug
+    prepare(float_model, inplace=True)
+    q_model.qconfig = qconfig_debug
+    prepare(q_model, inplace=True, white_list=white_list)
```
I recollect an issue where the qconfig for the quantized model is not over-written as it already has a qconfig. Does this work correctly?

hx89 replied:
Seems that's the issue for functionals, I will add function to remove the qconfig before preparing the model.

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -14,61 +16,93 @@ class EagerModeNumericSuiteTest(QuantizationTestCase):
-        for k, v in ob_dict.items():
-            self.assertTrue(v["float"].shape == v["quantized"].shape)
+
+    def test_compare_model_outputs(self):
```
Similar comments as previous PR. Please add a model with functional modules to verify if the API works with those.

raghuramank100 reviewed (changes requested):
Looks good, some suggestions

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -115,6 +118,22 @@ def forward(self, x, y):
             self.stats["float"] = torch.cat((self.stats["float"], y.detach()))
 
 
+class TensorLogger(Logger):
```
Can we reuse Record Observer here?

hx89 replied:
Record Observer inherits from observer, since we want to unify the shadow comparison and output comparison so that they can share same functions such as get_observer_dict() we need to inherit both from the same base Logger class.

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -242,3 +265,89 @@ def compare_model_stub(float_model, q_model, module_swap_list, data, Logger=Logg
+    return act_dict
+
+
+def remove_qconfig(module):
```
We need do this as part of convert, please add this logic there.

hx89 replied:
Seems if I remove qconfig in convert(), some tests related to QAT and fusion would fail: 

>   FAIL: 7
    caffe2/test:quantization - test_activations (quantization.test_quantize.TestQuantizationAwareTraining)
    caffe2/test:quantization - test_fuse_module_train (quantization.test_quantize.TestFusion)
    caffe2/test:quantization - test_conv_linear (quantization.test_quantize.TestQuantizationAwareTraining)
    caffe2/test:quantization - test_manual (quantization.test_quantize.TestQuantizationAwareTraining)
    caffe2/test:quantization - test_train_save_load_eval (quantization.test_quantize.TestQuantizationAwareTraining)
    caffe2/test:quantization - test_fusion_sequential_model_train (quantization.test_quantize.TestFusion)
    caffe2/test:quantization - test_compare_model_outputs (quantization.test_numeric_suite.TestEagerModeNumericSuite)

>

raghuramank100 replied:
I see, the problem is occuring because convert is called as part of prepare_qat. Please do the following:
1. Add remove_qconfig as an internal function _remove_qconfig in quantize.py
2. Please add a call to _remove_qconfig in quantize, quantize_qat and quantize_dynamic functions

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -242,3 +265,89 @@ def compare_model_stub(float_model, q_model, module_swap_list, data, Logger=Logg
+        q_module: the quantized module
+        white_list: list of module types to attach tensor logger
+    """
+    if white_list is None:
```
Instead, please make this a default argument:
``` python
def prepare_model_outputs(float_module, q_module, white_list=DEFAULT_NUMERIC_SUITE_COMPARE_MODEL_OUTPUT_WHITE_LIST, Logger= TensorLogger):
```

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -66,7 +66,7 @@ def propagate_qconfig_(module, qconfig_dict=None):
     if qconfig_dict is None:
         qconfig_dict = {}
-    _propagate_qconfig_helper(module, qconfig_dict)
+    _propagate_qconfig_helper(module, qconfig_dict, white_list)
```
Can you remove the if check on qconfig_dict and set the default to {}?

hx89 replied:
Seems set default to {} will result in lint error:
`[B006] Do not use mutable data structures for argument defaults. They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.`

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -122,7 +122,7 @@ def add_quant_dequant(module):
     return module
 
-def prepare(model, inplace=False):
+def prepare(model, inplace=False, white_list=None):
```
Can you set default argument for white_list=DEFAULT_QCONFIG_PROPAGATE_WHITE_LIST?

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -152,3 +154,48 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+        q_model = convert(q_model)
+        act_compare_dict = compare_model_outputs(model, q_model, data)
+        self.assertEqual(len(act_compare_dict), 7)
+        exptected_act_compare_dict_keys = {
```
Typo: expected

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -152,3 +154,48 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+            act_compare_dict = compare_model_outputs(float_model, q_model, data)
+            self.assertEqual(len(act_compare_dict), 2)
+            exptected_act_compare_dict_keys = {"conv.stats", "quant.stats"}
+            self.assertTrue(act_compare_dict.keys() == exptected_act_compare_dict_keys)
```
Typo: expected

raghuramank100 reviewed (changes requested):
Some more suggestions

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

hx89 commented:
Address comments.

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -160,3 +161,51 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
         for k, v in ob_dict.items():
             self.assertTrue(v["float"].shape == v["quantized"].shape)
+
+    @unittest.skipUnless(
```
Please ensure that the tests pass for QNNPACK also. See changes to existing tests (in test_numeric_suite) for information on how to do this.

hx89 replied:
Yes, I have added the same decorator as the existing tests have.

raghuramank100 reviewed (approved):
Looks good, please address remaining comments, thanks

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -227,6 +249,7 @@ def compare_model_stub(float_model, q_model, module_swap_list, data, Logger=Logg
     Args:
         float_module: the float module used to generate the q_module
         q_module: the quantized module
+        data: input data
```
Lets make the comments more descriptive
 module_swap_list: List of float module types at which shadow modules are attached. Shadow modules log the original
floating point module output along with the output of a quantized module for the same input.

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -218,7 +238,9 @@ def prepare_model_with_stubs(float_module, q_module, module_swap_list, Logger):
+    float_model, q_model, module_swap_list, data, Logger=RecordingLogger
+):
     r"""Returns a dict with key corresponding to module names and each entry being
     a dictionary with two keys 'float' and 'quantized', containing the output
```
Please add a more descriptive comment stating more details on what this function does. For example, look at the documentation for https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/conv.py#L214, so that the user gets more information. An example code snippet as part of the comments is very useful. You can reuse some material from the tutorial here too.

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -115,6 +118,22 @@ def forward(self, x, y):
             self.stats["float"] = torch.cat((self.stats["float"], y.detach()))
 
 
+class TensorLogger(Logger):
```
Couldnt leave a comment for RecordingLogger above: Why not keep the behavior of both Tensor and RecordingLoggers consistent. Both should not dequantize tensors, that way they just do simple capture of the data.

hx89 replied:
That's a good point, it's better to keep quantized tensor in both loggers.

raghuramank100 reviewed (changes requested):
Since there are a bunch of comments, chaning this to request changes so that we can iron out the remaining details.

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

hx89 commented:
Addressed comments.

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -278,6 +293,7 @@ def quantize_qat(model, run_fn, run_args, inplace=False):
     prepare_qat(model, inplace=True)
     run_fn(model, run_args)
     convert(model, inplace=True)
+    _remove_qconfig(model)
```
This will still cause errors, as convert (after training) will need a qconfig, so please remove it.

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -218,21 +244,128 @@ def prepare_model_with_stubs(float_module, q_module, module_swap_list, Logger):
     ob_dict = {}
     get_observer_dict(q_model, ob_dict, Logger)
     return ob_dict
+
```
NIT: Do we need to create an empty dict and pass it to get_observer_dict? Can we instead do:
ob_dict = get_observer_dict(q_model, Logger)?

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add module output comparison"


Add module output comparison API.

Differential Revision: [D21053197](https://our.internmc.facebook.com/intern/diff/D21053197/)

[ghstack-poisoned]
